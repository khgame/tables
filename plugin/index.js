
module.exports = {
  rows: require('./rows'),
  mark: require('./mark'),
  desc: require('./desc'),
  expand: require('./expand'),
  plain: require('./plain')
}
