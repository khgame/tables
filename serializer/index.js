const {
  readAndTranslate
} = require('../read')
const Path = require('path')
const fs = require('fs-extra')

/**
 * serialize files with selected serializers
 * @param {string} pathIn - input file path
 * @param {string} dirOut - output dir
 * @param {Object.<string,Serializer>} serializers
 */
function serialize (pathIn, dirOut, serializers) {
  let plugins = Object.values(serializers).reduce((prev, cur) => prev.concat(cur.plugins), [])
  // console.log('plugins', plugins)
  // console.dir(Object.values(serializers))
  let ret = readAndTranslate(pathIn, {
    plugins
  })
  for (let outName in serializers) {
    let fileContent = serializers[outName].file(ret, Path.parse(pathIn).name)
    fs.outputFileSync(Path.resolve(dirOut, outName), fileContent)
  }
}

module.exports = {
  serialize,
  jsonSerializer: require('./jsonSerializer').jsonSerializer,
  tsInterfaceSerializer: require('./tsInterfaceSerializer').tsInterfaceSerializer
}
