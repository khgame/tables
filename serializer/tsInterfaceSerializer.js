const Plugins = require('../plugin')
const _ = require('lodash')

const { getTypeName, supportedTypes } = require('../utils/schemaConvertor')

function getTsType (typeName) {
  switch (typeName) {
    case supportedTypes.String: return 'string'
    case supportedTypes.Float: return 'number'
    case supportedTypes.UFloat: return 'number'
    case supportedTypes.Int: return 'number'
    case supportedTypes.UInt: return 'number'
    case supportedTypes.Boolean: return 'boolean'
    case supportedTypes.Undefined: return 'undefined'
    case supportedTypes.Any: return 'any'
    default:
      if (typeName.startsWith(supportedTypes.Array)) {
        let typeInnerAll = getTypeName(typeName).args
        let typeInner = typeInnerAll.length > 0 ? typeInnerAll[0] : 'any'
        // console.log('typeInnerAll', 'array', typeName, typeInnerAll)
        return `${getTsType(typeInner)}[]`
      } else if (typeName.startsWith(supportedTypes.Pair)) {
        let typeInnerAll = getTypeName(typeName).args
        let typeInner = typeInnerAll.length > 0 ? typeInnerAll[0] : 'any'
        // console.log('typeInnerAll', 'Map', typeInnerAll)
        return `{key: string, val: ${getTsType(typeInner)}}`
      }
      return 'any'
  }
}

function dealSchema (schema, inArray = false, depth = 1) {
  let result = ''
  let space = ' '.repeat(depth * 2)
  let split = inArray ? ')|' : ';'
  let itemSql = []
  const rcv = (str) => {
    if (!inArray) {
      result += str
    } else {
      itemSql.push(str)
      // console.log('itemSql', depth, itemSql)
    }
  }

  for (let key in schema) {
    let temp = (inArray ? '(' : space + key + ': ')
    let schemaType = schema[key]
    if (_.isArray(schemaType)) {
      temp += dealSchema(schemaType, true, depth + 1)
      temp = temp.substr(0, temp.length - 1) + '[]' + split
      rcv(temp)
    } else if (_.isObject(schemaType)) {
      temp += '{\n'
      temp += dealSchema(schemaType, false, depth + 1)
      temp += space + '}' + split
      rcv(temp)
    } else {
      temp += `${getTsType(schemaType)}${split}`
      rcv(temp)
    }
    // if (inArray) break // nest array only accept the first type
    if (!inArray) {
      result += '\n'
    }
  }

  if (inArray) {
    let typeOr = []
    itemSql.forEach(str => {
      if (typeOr.indexOf(str) < 0) {
        // console.log('typeOr.indexOf(str)', typeOr, str, typeOr.indexOf(str))
        typeOr.push(str)
      }
    })
    let merge = typeOr.reduce((p, c) => p + c)
    // console.log('==> typeOr itemSql ', typeOr, itemSql, 'res', result, 'merge', merge)
    result += merge
  }
  return result
}

function culturelize (s) {
  s = s.toLowerCase()
  if (s.length <= 0) return ''
  let ret = s.substr(0, 1).toUpperCase() + s.substr(1)
  // console.log(ret)
  return ret
}

function makeInterfaceName (fileName) {
  return fileName.split(/[\s._-]/).reduce((p, c) => p + culturelize(c), 'I')
}

const tsInterfaceSerializer = {
  plugins: [Plugins.schema, Plugins.convert],
  file: (data, fileName) => {
    // console.log(data.schema)
    return `export interface ${makeInterfaceName(fileName)}{
${dealSchema(data.schema)}
}`
  }
}

module.exports = {
  dealSchema,
  makeInterfaceName,
  tsInterfaceSerializer
}
