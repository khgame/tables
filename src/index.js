export * from './plugin'
export * from './utils/read'
export * from './utils/names'
export * from './serializer'
