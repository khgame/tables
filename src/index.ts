export * from './plugin'
export * from './utils/read'
export * from './utils/names'
export * from './serializer'
export * from './core/protocol'
export * from './pipeline/layers'

export { exportTablesToTs } from './outOfBox/exportTablesToTs'
