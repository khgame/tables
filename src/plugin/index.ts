export * from './rows'
export * from './erows'
export * from './colMap'
export * from './mark'
export * from './desc'
export * from './expand'
export * from './plain'
export * from './convert'
export * from './schema'

