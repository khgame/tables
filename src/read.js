const XLSX = require('js-xlsx')

function readWorkBook (path) {
  // read the raw file
  return XLSX.readFile(path)
}

function translateWorkBook (workbook, sheetName) {
  // read the sheet with the given name or the first sheet
  const targetSheetInd = workbook.SheetNames.findIndex(n => n === (sheetName || '__data'))
  const targetSheetName = workbook.SheetNames[targetSheetInd > 0 ? targetSheetInd : 0]
  const sheet = workbook.Sheets[targetSheetName]

  // create return table
  let data = {}
  let colKeys = []

  // convert data types
  for (let key in sheet) {
    if (!sheet.hasOwnProperty(key)) continue
    let splitInd = -1
    for (let i = key.length - 1; i >= 0; i--) {
      if (key[i] >= '0' && key[i] <= '9') {
        continue
      }
      splitInd = i
      break
    }
    let col = key.slice(0, splitInd + 1)
    let row = key.slice(splitInd + 1)
    if (!row) { continue }
    if (!data[row]) data[row] = {}
    let { t, v, w } = sheet[key]
    data[row][col] = { t, v, w }
    colKeys[col] = 1
  }

  let cols = Object.keys(colKeys).sort((a, b) => a.length - b.length || a.localeCompare(b)) // 1. length 2. locale
  return {
    cols,
    data,
    getValue: (table_, row_, col_) => (!table_.data[row_]) || (!table_.data[row_][col_]) ? undefined : table_.data[row_][col_].v
  }
}

function readAndTranslate (path, options) {
  let table = translateWorkBook(readWorkBook(path), options.sheetName)
  if (options.plugins) {
    table = options.plugins.reduce((accumulate, plugin) => plugin(accumulate), table)
  }
  return table
}

module.exports = {
  readWorkBook,
  translateWorkBook,
  readAndTranslate
}
