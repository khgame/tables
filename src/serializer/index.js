const {
  readAndTranslate
} = require('../utils/read')
const Path = require('path')
const fs = require('fs-extra')

/**
 * serialize files with selected serializers
 * @param {string} pathIn - input file path
 * @param {string} dirOut - output dir
 * @param {Object.<string,Serializer>} serializers
 */
function serialize (pathIn, dirOut, serializers) {
  let plugins = Object.values(serializers).reduce((prev, cur) => prev.concat(cur.plugins), [])
  // console.log('plugins', plugins)
  // console.dir(Object.values(serializers))
  let ret = readAndTranslate(pathIn, {
    plugins
  })
  for (let outName in serializers) {
    let fileContent = serializers[outName].file(ret, Path.parse(pathIn).name)
    fs.outputFileSync(Path.resolve(dirOut, outName), fileContent)
  }
}

module.exports = {
  serialize,
  jsonSerializer: require('./jsonSerializer').jsonSerializer,
  jsSerializer: require('./jsSerializer').jsSerializer,
  tsSerializer: require('./tsSerializer').tsSerializer,
  tsInterfaceSerializer: require('./tsInterfaceSerializer').tsInterfaceSerializer
}
