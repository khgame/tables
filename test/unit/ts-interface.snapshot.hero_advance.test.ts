import { readAndTranslate } from '../../src/utils/read'
import { tableSchema, tableConvert } from '../../src/plugin'
import { tsInterfaceSerializer } from '../../src/serializer/tsInterfaceSerializer'
import * as Path from 'path'

describe('ts-interface snapshot (hero_advance.xlsx)', () => {
  it('matches snapshot', () => {
    const excel = Path.resolve(__dirname, '../../example/hero_advance.xlsx')
    const data = readAndTranslate(excel, { plugins: [tableSchema, tableConvert] })
    const code = tsInterfaceSerializer.file(data as any, 'hero_advance', '//imports', {})
    expect(code).toMatchSnapshot()
  })
})

