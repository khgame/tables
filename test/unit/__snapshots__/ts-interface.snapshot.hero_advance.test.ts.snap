// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ts-interface snapshot (hero_advance.xlsx) matches snapshot 1`] = `
"/** this file is auto generated */
//imports
        
export type HeroAdvanceTID = string & { readonly __HeroAdvanceTID: unique symbol };

export interface IHeroAdvance {
  system: number;
  id: number;
  star: number;
  str: number;
  agi: number;
  int: number;
  attr: Array<number>;
  type: "HERO_TYPE"|"temp";
}
"
`;
